src/disco/containers/InstallButton.js

```diff
@@ -0,0 +1,63 @@
+import React, { PropTypes } from 'react';
+import { connect } from 'react-redux';
+
+import { gettext as _ } from 'core/utils';
+
+import 'disco/css/InstallButton.scss';
+import {
+  DOWNLOADING,
+  INSTALLED,
+  UNKNOWN,
+  validInstallStates as validStates,
+} from 'disco/constants';
+
+export class InstallButton extends React.Component {
+  static propTypes = {
+    handleClick: PropTypes.func,
+    handleChange: PropTypes.func,
+    guid: PropTypes.string,
+    progress: PropTypes.number,
+    slug: PropTypes.string.isRequired,
+    state: PropTypes.oneOf(validStates),
+  }
+
+  static defaultProps = {
+    state: UNKNOWN,
+    progress: 0,
+  }
+
+  render() {
+    const { state, progress } = this.props;
+
+    if (validStates.indexOf(state) === -1) {
+      throw new Error('Invalid add-on state');
+    }
+
+    const isInstalled = state === INSTALLED;
+    const isDisabled = state === UNKNOWN;
+    const isDownloading = state === DOWNLOADING;
+    const switchClasses = `switch ${state}`;
+
+    return (
+      <div className={switchClasses} onClick={this.props.handleClick}
+        data-download-progress={isDownloading ? progress : 0}>
+        <input
+          className="visually-hidden"
+          checked={isInstalled}
+          disabled={isDisabled}
+          onChange={this.props.handleChange}
+          type="checkbox" />
+        <label>
+          {isDownloading ? <div className="progress"></div> : null}
+          <span className="visually-hidden">{_('Install')}</span>
+        </label>
+      </div>
+    );
+  }
+}
+
+function mapStateToProps(state, ownProps) {
+  return (state.installations || {})[ownProps.slug];
```

Comment:
I'm not sure why this `|| {}` is required but it explodes with `ownProps.slug is undefined` if it isn't there (on line 9 which is an import???).

Ideally `state.installations[ownProps.slug]` would always exist here because we would have already triggered a `dispatch({type: 'INSTALL_STATE`, payload: {slug: ownProps.slug, state: UNKNOWN}})` by now but I'm not sure of the right place to put that hook. Maybe in `componentWillMount()` but I'd have to test it.